File: pyredner_tensorflow/channels.py
import redner
import tensorflow as tf

class Channel:
    def __init__(self):
        self.radiance = redner.channels.radiance
        self.alpha = redner.channels.alpha
        self.depth = redner.channels.depth
        self.position = redner.channels.position
        self.geometry_normal = redner.channels.geometry_normal
        self.shading_normal = redner.channels.shading_normal
        self.uv = redner.channels.uv
        self.diffuse_reflectance = redner.channels.diffuse_reflectance
        self.specular_reflectance = redner.channels.specular_reflectance
        self.roughness = redner.channels.roughness
        self.generic_texture = redner.channels.generic_texture
        self.vertex_color = redner.channels.vertex_color
        self.shape_id = redner.channels.shape_id
        self.material_id = redner.channels.material_id

channels = Channel()

class RednerChannels:
    __channels = [
        redner.channels.radiance,
        redner.channels.alpha,
        redner.channels.depth,
        redner.channels.position,
        redner.channels.geometry_normal,
        redner.channels.shading_normal,
        redner.channels.uv,
        redner.channels.diffuse_reflectance,
        redner.channels.specular_reflectance,
        redner.channels.roughness,
        redner.channels.generic_texture,
        redner.channels.vertex_color,
        redner.channels.shape_id,
        redner.channels.material_id
    ]

    @staticmethod
    def asTensor(channel: redner.channels) -> tf.Tensor:
        assert isinstance(channel, redner.channels)

        for i in range(len(RednerChannels.__channels)):
            if RednerChannels.__channels[i] == channel:
                return tf.constant(i)

    @staticmethod
    def asChannel(index: tf.Tensor) -> redner.channels:
        try:
            channel = RednerChannels.__channels[index]
        except IndexError:
            print(f'{index} is out of range: [0, {len(RednerChannels.__channels)})')
            import sys
            sys.exit()
        else:
            return channel

class RednerCameraType:
    __cameratypes = [
        redner.CameraType.perspective,
        redner.CameraType.orthographic,
        redner.CameraType.fisheye,
    ]

    @staticmethod
    def asTensor(cameratype: redner.CameraType) -> tf.Tensor:
        assert isinstance(cameratype, redner.CameraType)

        for i in range(len(RednerCameraType.__cameratypes)):
            if RednerCameraType.__cameratypes[i] == cameratype:
                return tf.constant(i)


    @staticmethod
    def asCameraType(index: tf.Tensor) -> redner.CameraType:
        try:
            cameratype = RednerCameraType.__cameratypes[index]
        except IndexError:
            print(f'{index} is out of range: [0, {len(RednerCameraType.__cameratypes)})')
            import sys
            sys.exit()
        else:
            return cameratype

File: src/channels.cpp
#include "channels.h"
#include "cuda_utils.h"
#include <iostream>
#include <stdexcept>

ChannelInfo::ChannelInfo(const std::vector<Channels> &channels,
                         bool use_gpu,
                         int max_generic_texture_dimension) : use_gpu(use_gpu) {
    num_channels = (int)channels.size();
    radiance_dimension = -1;
    num_total_dimensions = compute_num_channels(channels, max_generic_texture_dimension);
    this->max_generic_texture_dimension = max_generic_texture_dimension;
    if (use_gpu) {
#ifdef __CUDACC__
        checkCuda(cudaMallocManaged(&this->channels, channels.size() * sizeof(Channels)));
#else
        assert(false);
#endif
    } else {
        this->channels = new Channels[channels.size()];
    }
    for (int i = 0; i < (int)channels.size(); i++) {
        if (channels[i] == Channels::radiance) {
            if (radiance_dimension != -1) {
                throw std::runtime_error("Duplicated radiance channel");
            }
            radiance_dimension = i;
        }
        if (use_gpu) {
            // We don't use unified memory to update here since another kernel might be running
#ifdef __CUDACC__
            checkCuda(cudaMemcpyAsync(&this->channels[i], &channels[i], sizeof(Channels), cudaMemcpyHostToDevice));
#else
            assert(false);
#endif
        } else {
            this->channels[i] = channels[i];
        }
    }
}

void ChannelInfo::free() {
    if (use_gpu) {
#ifdef __CUDACC__
        checkCuda(cudaFree(channels));
#else
        assert(false);
#endif
    } else {
        delete[] channels;
    }
}

int compute_num_channels(const std::vector<Channels> &channels,
                         int max_generic_texture_dimension) {
    int num_total_dimensions = 0;
    for (int i = 0; i < (int)channels.size(); i++) {
        switch(channels[i]) {
            case Channels::radiance: {
                num_total_dimensions += 3;
            } break;
            case Channels::alpha: {
                num_total_dimensions += 1;
            } break;
            case Channels::depth: {
                num_total_dimensions += 1;
            } break;
            case Channels::position: {
                num_total_dimensions += 3;
            } break;
            case Channels::geometry_normal: {
                num_total_dimensions += 3;
            } break;
            case Channels::shading_normal: {
                num_total_dimensions += 3;
            } break;
            case Channels::uv: {
                num_total_dimensions += 2;
            } break;
            case Channels::barycentric_coordinates: {
                num_total_dimensions += 2;
            } break;
            case Channels::diffuse_reflectance: {
                num_total_dimensions += 3;
            } break;
            case Channels::specular_reflectance: {
                num_total_dimensions += 3;
            } break;
            case Channels::roughness: {
                num_total_dimensions += 1;
            } break;
            case Channels::generic_texture: {
                num_total_dimensions += max_generic_texture_dimension;
            } break;
            case Channels::vertex_color: {
                num_total_dimensions += 3;
            } break;
            case Channels::shape_id: {
                num_total_dimensions += 1;
            } break;
            case Channels::triangle_id: {
                num_total_dimensions += 1;
            } break;
            case Channels::material_id: {
                num_total_dimensions += 1;
            } break;
            default: {
                assert(false);
            }
        }
    }
    return num_total_dimensions;
}
